### PR TITLE
Testing

### PR DESCRIPTION
**Pull request by rhkean**
migrate the .installed file from
    VARNAME val
to
    VARNAME=val
to remove need to "cat | grep | awk" parsing

also moved all variable definitions to top, then functions, then main loop at bottom

Could use some help testing.  (I've done a clean new install test with the v111 image)

Next:  will perform the same change for automation with dietpi.txt

### CODE REVIEW

**Comment by Fourdee** [comment]
@rhkean 
Looking good :+1: 

So the following will check for a variable named `SMBCLIENT` in memory/script, if it doesn't exist, sets to 0?

```
${SMBCLIENT:=0}
```

Cant we call `(( $SMBCLIENT == 1 ))` once the .installed file is loaded?

And this loads all the vars in the file to script/mem?

```
. /DietPi/dietpi/installed
```

As you have modified the cron job, we need to update the users system in the patch file (following does all DietPi cron jobs):

```
cp /DietPi/dietpi/conf/cron.daily_dietpi /etc/cron.daily/dietpi
chmod +x /etc/cron.daily/dietpi
cp /DietPi/dietpi/conf/cron.hourly_dietpi /etc/cron.hourly/dietpi
chmod +x /etc/cron.hourly/dietpi
```

I'll probably wait until you are finished with the automation side, then I'll merge and we can test it.

File: dietpi/dietpi-update [hunk 1]
```diff
@@ -67,7 +67,7 @@
 	Get_Server_Version(){
 
 		#Get server version file
-		curl -k -L https://raw.githubusercontent.com/Fourdee/DietPi/"$GITBRANCH"/dietpi/server_version > "$FILEPATH_TEMP"/server_version
+		curl -k -L https://raw.githubusercontent.com/rhkean/DietPi/"$GITBRANCH"/dietpi/server_version > "$FILEPATH_TEMP"/server_version
```
Comment (Fourdee) [inline dietpi/dietpi-update]: Dont forget to put these back before we merge :)

File: dietpi/dietpi-update [hunk 2]
```diff
@@ -124,7 +124,7 @@
 	Run_Update(){
 
 		#git clone Zip method (no need to install GIT)
-		curl -k -L https://github.com/Fourdee/DietPi/archive/"$GITBRANCH".zip > "$FILEPATH_TEMP"/update.zip
+		curl -k -L https://github.com/rhkean/DietPi/archive/"$GITBRANCH".zip > "$FILEPATH_TEMP"/update.zip
```
Comment (Fourdee) [inline dietpi/dietpi-update]: Dont forget to put these back before we merge :)

File: dietpi/patch_file [hunk 3]
```diff
@@ -601,6 +601,12 @@ _EOF_
 		echo -e "remove me" &> /dev/null
 		#-------------------------------------------------------------------------------
 
+	elif (( $VERSION_CURRENT == 117 )); then
```
Comment (Fourdee) [inline dietpi/patch_file:624]: This patch code should be in the v116 entry above. As the current version of the system will be 116 when the code is run, to update it to 117.

**Comment by Fourdee** [comment]
> Cant we call (( $SMBCLIENT == 1 )) once the .installed file is loaded?

Ok I did some testing and created a file with some vars.

```
BOB=2
JANE=3
DOE=2
STRING=whoop
```

Then I created a script, loaded the file as you did.

```
#!/bin/bash
. file
echo -e "$BOB"
if (( $BOB == 2 )); then
    echo yes
fi
```

```
root@DietPi:~# ./test
2
yes
```

So we can use the `(( $NAME == i ))` statements as before. 

Is there any other reason for changing all the if statements in dietpi-software to {} (`if (( ${RPIGPIO:=0} == 1 )); then`), or am i missing something? :)

**Comment by rhkean** [comment]
> Cant we call (( $SMBCLIENT == 1 )) once the .installed file is loaded?

yes.
you only need the ${VARIABLE:=0} variable expansion the first time you use a variable in the script.

it doesn't hurt to do that every time, so for the initial big-switch, I took the safer route.  I was going to go back and clean that up.

What I think is best to insure that each VARIABLE in .installed gets initialized correctly only once would be to change this block in the main loop:

```
#Generate blank .installed file. Doesnt exist on a fresh image.
if [ ! -f "$FP_INSTALLED_FILE" ]; then
    Write_InstallFileList
#load .installed file, update vars
else
    Read_InstallFileList
fi

```

To

```
#Generate blank .installed file. Doesnt exist on a fresh image.
if [ ! -f "$FP_INSTALLED_FILE" ]; then
    Write_InstallFileList
#load .installed file, update vars
else
    Read_InstallFileList
    Write_InstallFileList
fi

```

This will do 2 things:
1. properly initialized each variable at the biggining of the script, and insure that all other references can be just $VARNAME
2. make it so NEW variables only need to be added to the Write_InstallFileList function.  the logic being:
   1.  Read existing variables that are set by previous version, then write those values back to the file
   2.  AND write NEWVAR=${NEWVAR:=0} will initialize the new variable and add it to the .installed file.

File: dietpi/dietpi-update (hunk 2 above)
Comment (rhkean) [inline dietpi/dietpi-update]: DOH!  good catch.... sorry about that

File: dietpi/patch_file (hunk 3 above)
Comment (rhkean) [inline dietpi/patch_file:624]: oh... I put the "patching" code in the wrong block.

thank you

**Comment by Fourdee** [comment]
@rhkean 

> What I think is best to insure that each VARIABLE in .installed gets initialized correctly only once would be to change this block in the main loop:

Makes sense, go for it :+1: 

> yes.
> you only need the ${VARIABLE:=0} variable expansion the first time you use a variable in the script.
> it doesn't hurt to do that every time, so for the initial big-switch, I took the safer route. I was going to go back and clean that up.

Ah I see :)

Correct me if I'am wrong, but could we revert these back to `=0` (for the sake of consistency): https://github.com/rhkean/DietPi/blob/testing/dietpi/dietpi-software#L216-L325 as we are just declaring and initilizing the var which will be changed after `Read_InstallFileList` is read (if it exists)?

And for the if statements with var expansion (again, for the sake of consistency). Personally i'd like to revert these all back to a standard variable eg: `(( $VAR == i ))`. The var expansion seems unnecessary in these if statements.

Everything else looks spot on, going to have some fun testing this out :+1:

**Comment by rhkean** [comment]
> Correct me if I'am wrong, but we could revert these back to =0 (for the sake of consistency): https://github.com/rhkean/DietPi/blob/testing/dietpi/dietpi-software#L216-L325 as we are just declaring and initilizing the var which will be changed after Read_InstallFileList is read (if it exists).

no, these are the most critical to use variable expansion...
here's the logic (psuedo-code):

```
FP_INSTALLED_FILE="/DietPi/dietpi/.installed"
if [ -f "$FP_INSTALLED_FILE" ]; then
    . "$FP_INSTALLED_FILE"   # the "." is the same as "source"
fi
write_installed_file_with_variable_expansion()    #this call will basically do nothing unless an upgrade adds new variables to the file (patching file in patch_file no longer needed)
#
# do stuff that will probably change those values
#
write_installed_file_with_variable_expansion()    #this call will update values with new installs from this run of script
```

if we then add new variables in this 'write_installed_file_with_variable_expansion' function, it will always initialize our new apps with 0 (or, notice I initialized SSHSERVER_DROPBEAR to 2), but it won't change the values of what we read from an existing .installed file... the ${VARNAME:=0} will only set VARNAME equal to 0 if it doesn't already have a value.
Then, when it's called at the end of dietpi-software, it writes the new values....

so, basically,
1. source existing .installed file to get current values of installed apps
2. rewrite .installed file with just read values (and add new variables that may have been added due to upgrade)
3. perform installs and update variables to value of 2 as needed
4. at the end, write the installed file 1 more time with all the updated values

**Comment by rhkean** [comment]
I'll try make the changes and clean up the code tonight and push them to github so you can see what I mean

there's a giant :bulb: over my head, but I'm not sure that I'm communicating it well.

**Comment by xenfomation** [comment]
It makes perfect sense to me, especially this:

> f we then add new variables in this 'write_installed_file_with_variable_expansion' function, it will always initialize our new apps with 0 (or, notice I initialized SSHSERVER_DROPBEAR to 2), but it won't change the values of what we read from an existing .installed file... the ${VARNAME:=0} will only set VARNAME equal to 0 if it doesn't already have a value.

The second thing, @rhkean is keeping it consistent as you are doing to avoid false positives and in evaluations of the install, annoying unary operator issues because of the old school [ scalar (evaluation) value ]....  

I was knee deep in that last week playing with the code and other items outside of scope here!

If I can help or take care of the source/read file, take lines X-Y... let me know!  Good explanation, especially here:

> so, basically,
> 1. source existing .installed file to get current values of installed apps
> 2. rewrite .installed file with just read values (and add new variables that may have been added due to upgrade)
> 3. perform installs and update variables to value of 2 as needed
> 4. at the end, write the installed file 1 more time with all the updated values

-xenfomation | @xenfomation

**Comment by Fourdee** [comment]
@xenfomation @rhkean

Ok, but I still dont see why we need `${}` in the init stage of our code here: https://github.com/rhkean/DietPi/blob/testing/dietpi/dietpi-software#L216-L325

I did another test and its not required:

file

```
BOB=2
```

```
#!/bin/bash
BOB=0
. file
echo -e "$BOB"
if (( $BOB == 2 )); then
        echo -e "BOB = 2"
fi
```

Result

```
root@DietPi:~# ./test
2
BOB = 2
```

EDIT:
Now i see, completely missed it lol. I didn't see you moved the init code for those variables into another function. Makes perfect sense.
That will teach me to not rely on indent when scanning code :)

**Comment by Fourdee** [comment]
> there's a giant :bulb: over my head, but I'm not sure that I'm communicating it well.

@rhkean
Its probably more me not listening well, still trying to come over this lingering flu :( Sorry about that.

**Comment by Fourdee** [comment]
@rhkean

Just give us a nudge when this is ready to merge.

**Comment by rhkean** [comment]
try this:

filename:  demo-software.v1

```
#!/bin/bash
FP_INSTALLED_FILE=".installed"

write_installed(){
        cat << _EOF_ > "$FP_INSTALLED_FILE"
APPLICATION1=${APPLICATION1:=0}
APPLICATION2=${APPLICATION2:=0}
_EOF_
}

read_installed() {
        if [[ -f "$FP_INSTALLED_FILE" ]]; then
                . "$FP_INSTALLED_FILE"
        fi
}

fake_installation() {
        if [[ -n "$1" ]]; then
                if (( $1 == 1 )); then
                        APPLICATION1=2
                elif (( $1 == 2 )); then
                        APPLICATION2=2
                fi
        fi
}

### MAIN ###
read_installed
write_installed

fake_installation $1

write_installed
```

filename:  demo-software.v2

```
#!/bin/bash
FP_INSTALLED_FILE=".installed"

write_installed(){
        cat << _EOF_ > "$FP_INSTALLED_FILE"
APPLICATION1=${APPLICATION1:=0}
APPLICATION2=${APPLICATION2:=0}
APPLICATION3=${APPLICATION3:=0}
_EOF_
}

read_installed() {
        if [[ -f "$FP_INSTALLED_FILE" ]]; then
                . "$FP_INSTALLED_FILE"
        fi
}

fake_installation() {
        if [[ -n "$1" ]]; then
                if (( $1 == 1 )); then
                        APPLICATION1=2
                elif (( $1 == 2 )); then
                        APPLICATION2=2
                elif (( $1 == 3 )); then
                        APPLICATION3=2
                fi
        fi
}

### MAIN ###
read_installed
write_installed

fake_installation $1

write_installed
```

now, do this (note: no .installed file to begin with):

```
chmod +x demo-software.v*
./demo-software.v1

```

review .installed file

pretend to install an application:
`./demo-software.v1 2`
review .installed file

pretend that we upgraded from v1 to v2 (demo-upgrade has already run)
now, we're going to run demo-software.v2 for the first time (but not install anything)
`./demo-software.v2`
or run it and install the new "Application 3"
`./demo-software.v2 3`
review .installed file

**Comment by Fourdee** [comment]
@rhkean 
Awesome, works a charm :+1: . 
I take it from this you are planning to add the ability to run `dietpi-software PROGRAM` to set dietpi-software off to install that specific program? I've been meaning to do this for a while, but your updated code can help make this happen.

**Comment by rhkean** [comment]
well, I was just trying to demonstrate how the variable expansion logic was going to work....

but, I'll work on it.... not for V117, though....  I've changed too much of your logic for 1 version upgrade as it is... :smile:

**Comment by xenfomation** [comment]
Heheheh - I love these kind of scrums!!  I haven't read all of the recent posts, but the 'expansion' is good for two things:

${} can always preface a variable for easy lexical/eyeball sorting

So, instead of...
DIETPI="AWESOME"
.... only to be used later as echo "Hey, DietPi is absolutely $DIETPI"

We can...
${DIETPI:="DietPi is "}
${TRUTH:=" SO AMAZING"}
echo "${DIETPI}TRUTH"
... or echo "${DIETPIETRUTH}"

Headache, but not from this!

**Comment by Fourdee** [comment]
@rhkean 

>  I've changed too much of your logic for 1 version upgrade as it is... 

Hehe, aye, but its all good :+1: , much needed improvement to the code.

> ${DIETPI:="DietPi is "}
> ${TRUTH:=" SO AMAZING"}

Dam right :+1:

**Comment by Fourdee** [comment]
@rhkean 
Not to rush, but any idea when this might be ready to merge? I've been putting off `dietpi-software` work/additions until its done :)

**Comment by rhkean** [comment]
@Fourdee 
I rolled back my changes that were related to dietpi.txt for the time being, so I lost a day or so.  :cry: 
that file is used in too many different ways, so I decided to tackle it at a later point in time, after I've thought it through.

I'll clean up the changes related to the .installed file today.  It should be ready for you in the morning.

**Comment by Fourdee** [comment]
@rhkean 

> that file is used in too many different ways, so I decided to tackle it at a later point in time, after I've thought it through.

Yep, its kind of a mixbag of global variables, automation and 1st run settings. I will check `dietpi.txt` after the merge and see what needs changing.

> I'll clean up the changes related to the .installed file today. It should be ready for you in the morning.

Thanks Rob :+1:

**Comment by rhkean** [comment]
done

File: dietpi/conf/cron.hourly_dietpi [hunk 4]
```diff
@@ -28,13 +28,13 @@
 	#----------------------------------------------------------------
 	# Logging
 	# - clear all log files.
-	if (( $LOGGING_MODE == -1 )); then
+	if (( $INDEX_LOGGING_CURRENT == -1 )); then
 		/DietPi/dietpi/dietpi-logclear 1 &> /dev/null
 	# - Update current log files data to /"$HOME"/logfile_storage/*. Then clear.
-	elif (( $LOGGING_MODE == -2 )); then
+	if (( $INDEX_LOGGING_CURRENT == -2 )); then
```
Comment (Fourdee) [inline dietpi/conf/cron.hourly_dietpi]: Should  be `elif``

**Comment by Fourdee** [comment]
@rhkean 
Thanks. Still needs a few changes before we can merge:
- https://github.com/Fourdee/DietPi/pull/296/files#r61250202
- https://github.com/Fourdee/DietPi/pull/296/files#r60842897
- https://github.com/Fourdee/DietPi/pull/296/files#r60842902
- https://github.com/Fourdee/DietPi/pull/296/files#r60843428

EDIT: And this to patch file due to cronjob changes:

```
cp /DietPi/dietpi/conf/cron.daily_dietpi /etc/cron.daily/dietpi
chmod +x /etc/cron.daily/dietpi
cp /DietPi/dietpi/conf/cron.hourly_dietpi /etc/cron.hourly/dietpi
chmod +x /etc/cron.hourly/dietpi
```

**Comment by rhkean** [comment]
Think I got it all... :smiley:

**Comment by Fourdee** [comment]
@rhkean 
Excellent, thanks Rob :+1:  Merging now and i'll run some tests.

**Comment by Fourdee** [comment]
@rhkean 
`dietpi-software` is so much quicker at loading, great work :+1:

**Comment by Fourdee** [comment]
@rhkean 
I've added a function for loading `/DietPi/dietpi/.installed``in``dietpi-config``, and call it from the NAS menu. Wasn't refreshing installed state after installing NoIp etc: https://github.com/Fourdee/DietPi/commit/292fd6099ef56967b6b26d1278013f452c1e3c77

**Comment by rhkean** [comment]
thought I caught all of those situations.  sorry

**Comment by rhkean** [comment]
> dietpi-software is so much quicker at loading, great work
> Awesome!
> I'm glad the changes had a positive impact.

**Comment by Fourdee** [comment]
@rhkean 

> thought I caught all of those situations. sorry

No worries, that what testing is for :)

**Comment by Fourdee** [comment]
@rhkean 
Been testing this over the last few days, solid as a rock. Did a automation installation today, had to modify the code slightly to work with `=` in `.installed_file`: https://github.com/Fourdee/DietPi/commit/93261f78895fec7be9fd704c8bc8f9a45ec65431.
Also removed all references to `DietPi_Software_Addition()` in the patchfile.

Great work on this Rob :+1: .

**Comment by rhkean** [comment]
> Also removed all references to DietPi_Software_Addition() in the patchfile.

ha!
I had done that, but put it back, because I wasn't 100% sure how the update worked.
- did it update 1 version at a time 
- or jump straight from the img version to the latest 

and there were too many other moving parts to risk it at the time.

Seriously @Fourdee , I am really glad that my changes have made a positive impact on your wonderful creation.  :smiley:

**Comment by Fourdee** [comment]
@rhkean 
The update system works in this order:
- Obtains the latest dietpi scripts (/DietPi/dietpi)
- Runs through each patch and number, 1 at a time until it reaches the latest version.

So, runs patch 115, then patch 116 etc etc. As your updated the `dietpi-software` script to handle new items on its own, we dont need to add them into the patch file anymore. Only the automation entry in `dietpi.txt` is needed.

**Comment by Fourdee** [comment]
@rhkean 
Found a bug when `/DietPi/dietpi/.installed` does not exist (eg: 1st run).

```
    #--------------------------------------------------------------------------------------
    #load .installed file, update vars, if it exists
    Read_InstallFileList

    #Generate blank .installed file if it doesnt exist on a fresh image or update with current values.
    Write_InstallFileList
echo -e "1 $USBDRIVE"
echo -e "2 $NETDATA"
```

Example variables `USBDRIVE` and `NETDATA` have not been initialized, causes:

```
/DietPi/dietpi/dietpi-software: line 4718: ((: == 0 : syntax error: operand expected (error token is "== 0 ")
/DietPi/dietpi/dietpi-software: line 4836: ((: == 2 : syntax error: operand expected (error token is "== 2 ")
```

Fix is to reload file after writing, so:

```
    #load .installed file, update vars, if it exists
    Read_InstallFileList

    #Generate blank .installed file if it doesnt exist on a fresh image or update with current values.
    Write_InstallFileList

    #++ load .installed file again to update vars.
    Read_InstallFileList

```

**Comment by rhkean** [comment]
That shouldn't be required...

```
echo "VARIABLE=${VARIABLE:=0}"
if (( $VARIABLE == 0 )); then
    echo "variable initialized to 0"
fi
```

should output

```
VARIABLE=0
variable initialized to 0
```

weird

**Comment by Fourdee** [comment]
@rhkean 

The only time the variables are initialized is when `Read_InstallFileList` is called.

As the .installed file does not exist on 1st run:

```
if [[ -f "$FP_INSTALLED_FILE" ]]; then
    . "$FP_INSTALLED_FILE"
fi
```

The vars do not get initialized. So we need to `Read_InstallFileList` after `Write_InstallFileList` to init the vars after a `.installed` file is generated

**Comment by rhkean** [comment]
no.... 
the := does the initialization

here:
1.  echo $VAR
2.  echo ${VAR}  #==> same as # 1
3.  echo ${VAR:-0}
   equivalent to:

```
if [ -z "$VAR"]; then
    echo 0
else
    echo $VAR
fi
```
1.  echo ${VAR:=0}  (this is the one we're using
   equivalent to:

```
if [-z "$VAR"]; then
    VAR=0
fi
echo $VAR
```

**Comment by rhkean** [comment]
Here's a good explanation:
http://www.tldp.org/LDP/abs/html/parameter-substitution.html

**Comment by rhkean** [comment]
then there's THIS test....  I'm at a loss...

```
#!/bin/bash

FP_INSTALLED_FILE="testfile.txt"

# this if block just wipes out the test file for repeat runs of this test 
# to insure that testfile.txt doesn't exist 
if [[ -f "$FP_INSTALLED_FILE" ]]; then
        rm "$FP_INSTALLED_FILE"
fi

Read_File(){
        if [[ -f "$FP_INSTALLED_FILE" ]]; then
                . "$FP_INSTALLED_FILE"
        fi
}

Write_File(){
        cat << _EOF_ > "$FP_INSTALLED_FILE"

#DietPi External Drive
USBDRIVE=${USBDRIVE:=0}

#Extra test var
TESTVAR=${TESTVAR:=0}

_EOF_

}

Read_File
Write_File

# ODDLY, this works
if [[ $USBDRIVE -eq 0 ]]; then
        echo "USBDRIVE is 0"
else
        echo "USBDRIVE is not 0"
fi

# BUT, this errors
if (( $USBDRIVE == 0 )); then
        echo "USBDRIVE is 0"
else
        echo "USBDRIVE is not 0"
fi

```

**Comment by rhkean** [comment]
<<< previous post deleted after further testing >>>

**Comment by rhkean** [comment]
hmmm

I think I figured out why it's not working

I think the 
`cat << _EOF_ > "$FILE"
...
_EOF_`
is spawning a new process
because, if I replace the cat with a series of echo's
it works
doesn't accomplish what we need it to....
but it does the variable assignments like it's supposed to

**Comment by rhkean** [comment]
```
#!/bin/bash

FP_INSTALLED_FILE="testfile.txt"

if [[ -f "$FP_INSTALLED_FILE" ]]; then
        rm "$FP_INSTALLED_FILE"
fi

Read_File(){
        if [[ -f "$FP_INSTALLED_FILE" ]]; then
                . "$FP_INSTALLED_FILE"
        fi
}

Write_File(){
####    with the cat to output, I believe it's spawning another process
####    that is by we're losing the initialization of the variables.

#       cat << _EOF_ > "$FP_INSTALLED_FILE"
#
##DietPi External Drive
#USBDRIVE=${USBDRIVE:=0}
#
##Extra test var
#TESTVAR=${TESTVAR:=0}
#
#_EOF_

#### This successfully initializes the variables
#### Very frustrating

        echo USBDRIVE=${USBDRIVE:=0}
        echo TESTVAR=${TESTVAR:=0}

}

Read_File
Write_File

if (( $USBDRIVE == 0 )); then
        echo "USBDRIVE is 0"
else
        echo "USBDRIVE is not 0"
fi

```

**Comment by Fourdee** [comment]
@rhkean 

> I think the 
> cat << _EOF_ > "$FILE" ... _EOF_
> is spawning a new process
> because, if I replace the cat with a series of echo's
> it works
> doesn't accomplish what we need it to....
> but it does the variable assignments like it's supposed to

That explains it, good find :+1:. I think its also because we are redirecting cat output to a file, so the vars dont get loaded into memory?

**Comment by rhkean** [comment]
yeah... working on a resolution...

the read-if-exists, write, read-if-exists works, but it bugs me

:smile:

**Comment by rhkean** [comment]
fixed it!!!!

resolved with pull request #304
set a multi-line variable
echo that variable (quoted, to preserve whitespace) to the output file 

```
#!/bin/bash

FP_INSTALLED_FILE="testfile.txt"

Read_File(){
        if [[ -f "$FP_INSTALLED_FILE" ]]; then
                . "$FP_INSTALLED_FILE"
        fi
}

Write_File(){
local output="
#DietPi External Drive
USBDRIVE=${USBDRIVE:=0}

#Extra test var
TESTVAR=${TESTVAR:=0}

"
        echo "$output" > $FP_INSTALLED_FILE
}

Read_File
Write_File

if (( $USBDRIVE == 0 )); then
        echo "test pass"
else
        echo "test FAIL"
fi
```

**Comment by Fourdee** [comment]
@rhkean 
Legend :+1: . Great fix.

**Comment by rhkean** [comment]
hey..... that 50 micro-seconds was important to me!!!  :wink:

**Comment by rhkean** [comment]
AMAZINGLY helpful post about bash redirection...

http://www.catonmat.net/blog/bash-one-liners-explained-part-three/

I'm going to have to read the rest of his blogs....
